### PR TITLE
Fix handling of -Path when it literally contains wildcard char

### PR DESCRIPTION
**Pull request by kwkam**
Fix handling of `-Path` argument that contains wildcard char (eg. ```./`[file`].txt```) on some Cmdlet.
Fix tab completion when path contains wildcard char (eg. ```./`[fi```[TAB] did not get completed).
Make `Get-Command` try to resolve `-Path` as literal when it contains invalid wildcard pattern.

I believe this also solve/work around the following issues:
#3724 #3725 #4726 #6232

## PR Summary

_*Microsoft.PowerShell.Commands.Management/commands/management/Navigation:_
**Move-Item**: Set the `context.SuppressWildcardExpansion` in `MoveItem` instead of escaping the path every time. Also solve the issue where Move-Item complains -Path wildcard pattern is not valid when -Path contains special chars that forms an invalid pattern.
**Rename-Item**: Unescape non-literal, non-glob path in `ProccessRecord` and set the `context.SuppressWildcardExpansion` in `RenameItem`. This solve the issue where Rename-Item complains -Path does not exist when both -Path and CWD contains special char.

_*System.Management.Automation/engine/CommandCompletion/CompletionCompleters:_
Use `WildcardPattern::Escape` to escape completion text of filename.

_*System.Management.Automation/engine/CommandSearcher:_
Do not resolve path as glob when it contains invalid wildcard pattern (eg. `./[.ps1`).
Merge the path resolving code from `GetNextFromPath` into `ResolvePSPath` and use that function.

_*System.Management.Automation/engine/GetCommandCommand:_
Do not treat path as pattern when it contains invalid wildcard pattern (eg. `./[.ps1`).

_*System.Management.Automation/engine/SessionStateContainer:_
**Get-ChildItem**: Unescape non-literal, non-glob path before existence checking. This solve the issue where Get-GhildItem does not behave correctly when -Path contains special char. For example, ```Get-ChildItem -Path './`[dir`]'``` will complain `Cannot find path ...` while ```Get-ChildItem -Path './`[dir`]' -Depth 0``` will work fine.
**New-Item**: Unescape non-literal path when -Name is not set. This works around for New-Item treating -Path as literal path while it can also be globbable. For example, assuming there is `[file]1` in current directory, and tab completion suggests ```./`[file`]1``` for the -Path argument, but ```New-Item -Path './`[file`]2'``` will create a file named ``` `[file`]2``` instead of `[file]2`.

_*System.Management.Automation/engine/regex:_
`WildcardPattern::Escape` should also escape ``` ` ```, since `WildcardPattern::Unescape` would unescape it, and the matcher parse it as escape char.

_*System.Management.Automation/namespaces/LocationGlobber:_
Do not escape CWD when LiteralPath is used. This causes issues when both -LiteralPath and CWD contains special char.
Do not pass regex escaped string to `WildcardPattern::Get`. This fails tab completion when doing ```./path/to/`[f```[TAB].

_*System.Management.Automation/utils/PathUtils:_
Unescape non-literal, non-glob path before calling `GetUnresolvedProviderPathFromPSPath` since it only accepts literal path. This solve the issue where some commands depending on that method will treat -Path as literal unintentionally. For example, ```Out-File -Path './`[out`].txt' -Append``` will create a new file named ``` `[out`].txt``` instead of writing to ```[out].txt```.

## PR Checklist

Note: Please mark anything not applicable to this PR `NA`.

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [x] Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] User facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [ ] Issue filed - Issue link:
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
    - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

### CODE REVIEW

**Comment by adityapatwardhan** [comment]
@kwkam Please add tests. Marking the PR as `WIP`

**Comment by iSazonov** [comment]
@kwkam Could you continue with the PR?

**Comment by kwkam** [comment]
@adityapatwardhan Done. I am not sure whether they are in the proper file though

**Comment by adityapatwardhan** [comment]
@daxian-dbw @lzybkr @anmenaga Can you have a look?

**Comment by adityapatwardhan** [comment]
@anmenaga @daxian-dbw ping ...

**Comment by kwkam** [comment]
I would like to know, can I rebase this to make the commit history cleaner/more readable?

**Comment by iSazonov** [comment]
Please don't rebase until a maintainer ask you.

**Comment by kwkam** [comment]
A question about the `Get-Command` cmdlet.
https://github.com/PowerShell/PowerShell/blob/be109922f36d509d7bf512a49ca91dab1ef3100c/test/powershell/Modules/Microsoft.PowerShell.Core/Pester.Commands.Cmdlets.GetCommand.Tests.ps1#L6-L50
According to the test file, `Get-Command` seems to be designed to handle both wildcard and literal path with a single `-Name` argument, but when the filename is `[.ps1`, it will not be able to handle it, throwing an exception about the invalid pattern: `The specified wildcard character pattern is not valid: [.ps1`.

How should this be resolved? Catch the exception and fallback to use literal path,
or, add another argument to `Get-Command` to handle wildcard path separately,
or, do not allow literal path for command (has to be executed as ```& './`[.ps1'```),
or, some other better solution or exclude the issue from this PR?

**Comment by adityapatwardhan** [comment]
@daxian-dbw Can you have a look?

**Comment by adityapatwardhan** [comment]
@kwkam Please have a look at the test failure. 

Please also at `[Feature]` in the last commit to run all the tests in CI.  https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr

**Comment by kwkam** [comment]
@adityapatwardhan Test failures related to the changes are solved (I have no idea why the remote session test is failing). Please let me know if anything were missing/wrong.

**Comment by adityapatwardhan** [comment]
@anmenaga  @daxian-dbw ping..

File: test/powershell/Modules/Microsoft.PowerShell.Management/Remove-Item.Tests.ps1 [hunk 1]
```diff
@@ -109,6 +112,14 @@ Describe "Remove-Item" -Tags "CI" {
...
 	    Test-Path (Join-Path -Path $testpath -ChildPath file2.wav) | Should -BeFalse
 	}
+
+        It "Should be able to remove file when path contains special char" {
+            New-Item -Path $testpath -Name $testfileSpName -ItemType File -Force
+            $testfilepathSp | Should -Exist
+
+            Remove-Item -Path $testfilepathSp
+            $testfilepathSp | Should -Not -Exist
+        }
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Remove-Item.Tests.ps1]: Please correct a indentation of the It block.

Comment (kwkam) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Remove-Item.Tests.ps1]: I believe my indentation is correct.
What I notice is, there are 93 lines of code that is mixing tab/space for indentation.
Should I fix them here?

Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Remove-Item.Tests.ps1]: Sorry, GitHub show a shift.
Closed.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Pester.Commands.Cmdlets.GetCommand.Tests.ps1 [hunk 2]
```diff
@@ -46,6 +48,11 @@ Describe "Tests Get-Command with relative paths and wildcards" -Tag "CI" {
         $result | Should -Not -BeNullOrEmpty
         $result | Should -Be WildCardCommand[B].exe
 
+        # This should find the file [.exe
+        $result = Get-Command -Name .\[.exe -Type Application
+        $result | Should -Not -BeNullOrEmpty
+        $result | Should -Be [.exe
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Pester.Commands.Cmdlets.GetCommand.Tests.ps1]: Please use -BeExactly for string.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs [hunk 3]
```diff
@@ -3308,173 +3308,165 @@ protected override void ProcessRecord()
...
                         string resolvedPath = resolvedPathInfo.Path;
-                        MoveItem(resolvedPath);
+                        MoveItem(resolvedPath, literalPath: true);
                     }
                 }
             }
         } // ProcessRecord
 
-        private void MoveItem(string path)
+        private void MoveItem(string path, bool literalPath)
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs]: Should we assign default bool literalPath = false?

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs [hunk 4]
```diff
@@ -3607,22 +3599,101 @@ protected override bool ProviderSupportsShouldProcess
...
+                    RenameItem(Path, literalPath: true);
+                    continue;
+                }
+
+                Collection<PathInfo> resolvedPaths = GetResolvedPaths(Path);
+
+                switch (resolvedPaths.Count)
+                {
+                    case 0:
+                        // Do nothing
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs]: Please remove obvious comment.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs [hunk 5]
```diff
@@ -3607,22 +3599,101 @@ protected override bool ProviderSupportsShouldProcess
...
+                                Path);
+                        WriteError(
+                            new ErrorRecord(
+                                invalidOperation.ErrorRecord,
+                                invalidOperation));
+                        break;
+                }
+            } while (false);
+
+        } // ProcessRecord
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs]: Please remove the comment.

File: src/System.Management.Automation/engine/CommandSearcher.cs [hunk 6]
```diff
@@ -458,84 +458,21 @@ private CommandInfo GetNextFromPath()
...
-                    File.Exists(resolvedPaths[0]))
-                {
-                    string path = resolvedPaths[0];
+            CommandDiscovery.discoveryTracer.WriteLine(
+                "Trying to resolve the path as an PSPath");
 
-                    CommandDiscovery.discoveryTracer.WriteLine(
-                        "Path resolved to: {0}",
-                        path);
+            // Find the match if it is.
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandSearcher.cs]: Seems previous tracer comment is more useful. Please remove comment in the line or reword.

File: src/System.Management.Automation/engine/CommandSearcher.cs [hunk 7]
```diff
@@ -458,84 +458,21 @@ private CommandInfo GetNextFromPath()
...
+            string resolvedPath = ResolvePSPath(_commandName);
 
-                    result = GetInfoFromPath(path);
-                }
-            } while (false);
+            // If the path was resolved, and it exists
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandSearcher.cs]: Please remove obvious comment.

**Review by iSazonov — COMMENTED** [review]
@kwkam The PR is very large. Can we split it on some new PRs?

File: src/Microsoft.PowerShell.Commands.Management/resources/NavigationResources.resx [hunk 8]
```diff
@@ -168,6 +168,9 @@
   <data name="MoveItemInUse" xml:space="preserve">
     <value>Cannot move item because the item at '{0}' is in use.</value>
   </data>
+  <data name="RenameMultipleItemError" xml:space="preserve">
+      <value>Cannot rename item because the path '{0}' resolved to multiple items. Only one item can be renamed at a time.</value>
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Management/resources/NavigationResources.resx]: @SteveL-MSFT Could you please review the resource string?

File: test/powershell/Modules/Microsoft.PowerShell.Management/Remove-Item.Tests.ps1 (hunk 1 above)
Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Remove-Item.Tests.ps1]: I hope we already fixed most of tabs in test .ps1 files. 
Feel free to fix remains in new PR.

**Comment by iSazonov** [comment]
@kwkam The PR is very large. Could you split it on some small PRs for more fast review?

**Comment by kwkam** [comment]
@iSazonov Sorry, I am afraid I would not have free time get it done very soon. Feel free to fork this if you are in hurry.
P.S. Appending `?w=1` to the URL (or `--ignore-white-space` on command line) could help a bit when viewing the diff.

**Pull request closed without merging**